### PR TITLE
[dvdplayer] Prioritize number of channels over codec over default flag.

### PR DESCRIPTION
**Pull request by ace20022**
@MartijnKaijser especially for you ;) Trac: http://trac.kodi.tv/ticket/14697
@FernetMenta (http://forum.kodi.tv/showthread.php?tid=192280)

### CODE REVIEW

**Comment by FernetMenta** [comment]
great, this was fast :)
maybe we should make this optional by a setting? the average system won't benefit from it. decoding DTS-HD MA and output it as stereo just puts more load on the CPU.

**Comment by Montellese** [comment]
I agree with @FernetMenta. The default behaviour should be to consider the default flag before the language (if no preferred language has been set by the user), the number of channels or the codec. Then the user can specify a preferred language and whether he wants to follow the default flag or prefer the "better" audio track.

**Comment by Voyager1** [comment]
Why would we need for a setting, since you already know the AV receiver's capabilities. Ideally, you would only prefer a DTS track over a regular stereo track if the receiver can actually decode it. Same for DTS-HD MA, Dolby TrueHD and so on...

**Comment by ace20022** [comment]
I also thought about making it fully customizable in the expert setting, also for subs. In a way my addon does it for subs. But that's rather a long-term goal and I coule need some help on the gui side.
@Voyager1 I guess a dts track decoded to pcm is better than a stereo track for example.

**Comment by Voyager1** [comment]
@ace20022 you're right, only if multichannel PCM via HDMI is available. For those (like me) who have an SPDIF connection to an older 5.1 (non-HD audio formats capable) receiver you would have to re-code it to AC3...

**Comment by FernetMenta** [comment]
Note that we can't query audio capabilities for this purpose because it is unreliable.

**Comment by Voyager1** [comment]
Why unreliable? Not talking about the actual AE device, but about the audio settings "as configured": number of (LPCM) channels, passthrough, AC3 capable, EAC3 capable, DTS capable, TrueHD capable, etc. etc. With that information it should be possible to determine whether a "better" audio track will actually be better or not.

**Comment by ace20022** [comment]
@FernetMenta @Montellese added a setting. Could you please check if the setting and it's description are understandable?

File: system/settings/settings.xml [hunk 1]
```diff
@@ -393,6 +393,11 @@
           </constraints>
           <control type="list" format="string" />
         </setting>
+        <setting id="videoplayer.preferdefaultflag" type="boolean" label="37040" help="37041">
+          <level>1</level>
+          <default>false</default>
```
Comment (FernetMenta) [inline system/settings/settings.xml]: the default needs to be true.

**Comment by ace20022** [comment]
done

**Comment by FernetMenta** [comment]
looks ok to me

@Voyager1 your suggestion with audio settings may work but I think it is much clearer having a dedicated setting for this.

**Comment by Jalle19** [comment]
I agree with having a setting for this since it's non-standard behavior.

**Comment by MartijnKaijser** [comment]
Awesome :) 
Can't wait to try it out

**Comment by fritsch** [comment]
jenkins build this please

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
build errors are not related
@ace20022 ready to merge?

**Comment by ace20022** [comment]
We are not in the feature merge window, or was this abolished?

**Comment by FernetMenta** [comment]
after Gotham release we suspended merge window process for a while to get PRs in which piled up during bugfixing on release. I was thinking that this is also true now

**Comment by fritsch** [comment]
I am thinking the same, all in with it :-)

File: language/English/strings.po [hunk 2]
```diff
@@ -16179,7 +16179,19 @@ msgctxt "#37039"
...
+#. Setting #37040 Settings -> Video -> Playback
+#: system/settings/settings.xml
+msgctxt "#37040"
+msgid "Prefer default audio streams"
+msgstr ""
+
+#. Description of setting #37040 Settings -> Video -> Playback
+#: system/settings/settings.xml
+msgctxt "#37041"
+msgid "Use audio streams that are flagged as "default" (and match the preferred language) regardless of the number of channles or the used codec."
```
Comment (Voyager1) [inline language/English/strings.po]: sorry for noticing typo late: should be "channels" instead of "channles"

Comment (ace20022) [inline language/English/strings.po]: lol, thx!

**Comment by da-anda** [comment]
If I may add a partly related note here. Is there a way to priorize 2ch stereo over 2ch-AC3? Reason is, with a SPDIF connection you can't controll the volume on AC3 passthrough from within Kodi while 2ch-stereo ofc would work. So expecially when watching LiveTV I'd prefer 2ch-stereo over 2ch-AC3. Not sure how such a behavior could be configured.

**Comment by FernetMenta** [comment]
comparing apples with pears :)
you compare stereo with AC3? stereo is no codec.

**Comment by da-anda** [comment]
I'm not comparing - I was asking how and if such a priorization could be handled. In LiveTV I really don't care about a maybe better bitrate in 2ch-AC3 - I'd like to keep the comfort of changing the volume inside Kodi and let the AVR handle surround upmix

**Comment by FernetMenta** [comment]
that's a question of passthrough or non-passthrough audio but not dependent on the selected stream. Even for 2channel stereo I would prefer the AC3 track over the mpeg one.

**Comment by Voyager1** [comment]
the whole codec prioritization thing is a bit arbitrary based on 'assumed' audio quality. It would be nice that instead of the arbitrary order (as in ```StreamUtils::GetCodecPriority``` https://github.com/xbmc/xbmc/blob/master/xbmc/utils/StreamUtils.cpp ) one could simply define it in the settings (the easiest would be AdvancedSettings: you could assign the values returned in that function to whatever you prefer - ranking DTS over FLAC over TrueHD for instance)

**Comment by FernetMenta** [comment]
I doubt that there is much material out there having i.e. a DTS-HD and a TrueHD track.

**Comment by ace20022** [comment]
@Voyager1 
> I also thought about making it fully customizable in the expert setting, also for subs. In a way my addon does it for subs. But that's rather a long-term goal and I could need some help on the gui side.

@FernetMenta Probably not even on a (official) demo disc ;)

But I also wondered why flac is the top element. And flac together with DTS-HD, hmm

**Pull request merged**